### PR TITLE
KNOX-2571 - Introducing Knox Home page profiles

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

Added the ability to declare an arbitrary number of homepage profiles in `gateway-site.xml`. The relevant configuration prefix is `knox.homepage.profile.` which should be finished with the profile name (e.g. `knox.homepage.profile.smolnar` declares the profile called `smolnar`).
The purpose of these profiles is to
- control what's going to be displayed in General Proxy Information
- optionally, filter topologies too

The following language elements should be used when declaring a profile in `gateway-site.xml`
- gpi_version - allows `Knox Version`
- gpi_cert - allows `TLS Public Certificate`
- gpi_admin_ui - allows `Admin UI URL`
- gpi_admin_api - allows `Admin API Details`
- gpi_md_api - allows `Metadata API`
- gpi_tokens - allows `Integration Tokens`
- top_$TOPOLOGY_LIST - allows to define a comma-separated list of topology names that should be included in the `Topologies` section

Let's see a sample:
```
    <property>
        <name>knox.homepage.profile.smolnar</name>
        <value>gpi_version,gpi_tokens,top_other</value>
    </property>
```
The `smolnar` profile will allow displaying the `Knox Version` and `Integration Tokens` elements in `General Proxy Information` and will filter topologies such as only the topology called `other` will be displayed in `Topologies`.

Other than adding homepage profile support I also moved the session information block (username and optional logout button) into the top-left corner.

## How was this patch tested?

Manually tested (I used the above described `smolnar` profile during my tests):
Screen Shot 2021-04-07 at 2 41 43 PM
Screen Shot 2021-04-07 at 2 41 57 PM
Screen Shot 2021-04-07 at 2 42 12 PM
Screen Shot 2021-04-07 at 2 42 35 PM
Screen Shot 2021-04-07 at 2 43 23 PM

```
curl -ku admin:admin-password https://localhost:8443/gateway/metadata/api/v1/metadata/profiles/full
{
  "gpi_tokens" : "true",
  "gpi_version" : "true",
  "gpi_md_api" : "true",
  "gpi_cert" : "true",
  "gpi_admin_ui" : "true",
  "gpi_admin_api" : "true",
  "topologies" : ""
}


curl -ku admin:admin-password https://localhost:8443/gateway/metadata/api/v1/metadata/profiles/smolnar
{
  "gpi_tokens" : "true",
  "gpi_version" : "true",
  "gpi_md_api" : "false",
  "gpi_cert" : "false",
  "gpi_admin_ui" : "false",
  "gpi_admin_api" : "false",
  "topologies" : "other"
}
```

### CODE REVIEW

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java [hunk 1]
```diff
@@ -1209,4 +1211,23 @@ public long getKeystoreCacheEntryTimeToLiveInMinutes() {
...
+      if (config.getKey().startsWith(KNOX_HOMEPAGE_PROFILE_PREFIX)) {
+        profiles.put(config.getKey().substring(KNOX_HOMEPAGE_PROFILE_PREFIX.length()), getTrimmedStringCollection(config.getKey()));
+      }
+    });
+    return profiles;
+  }
+
+  private Map<String, Collection<String>> getPreConfiguredProfiles() {
+    final Map<String, Collection<String>> profiles = new HashMap<>();
+    profiles.put("full", Arrays.asList("gpi_version", "gpi_cert", "gpi_admin_ui", "gpi_admin_api", "gpi_md_api", "gpi_tokens"));
```
Comment (moresandeep) [inline gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java:1228]: Why not define these as static constants? added and reduced when needed (by cloning)

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java [hunk 2]
```diff
@@ -1209,4 +1211,23 @@ public long getKeystoreCacheEntryTimeToLiveInMinutes() {
   public boolean isGatewayServerIncomingXForwardedSupportEnabled() {
     return getBoolean(KNOX_INCOMING_XFORWARDED_ENABLED, true);
   }
+
+  @Override
+  public Map<String, Collection<String>> getHomePageProfiles() {
+    final Map<String, Collection<String>> profiles = getPreConfiguredProfiles(); // pre-configured profiles might be overwritten
+    this.forEach(config -> {
```
Comment (moresandeep) [inline gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java:1218]: i don't follow what `this` object is here

**Review by moresandeep — APPROVED** [review]

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java (hunk 2 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java:1218]: This class implements `Iterable`

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java (hunk 1 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java:1228]: Yepp; I'm going to figure out how to get rid of circular dependency and submit a new patch.

[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java:1228]: Yepp; I'm going to figure out how to get rid of circular dependency and submit a new patch.

File: gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/KnoxMetadataResource.java [hunk 3]
```diff
@@ -250,4 +253,18 @@ private ServiceModel getServiceModel(HttpServletRequest request, String gatewayP
     return serviceModel;
   }
 
+  @GET
+  @Produces({ APPLICATION_JSON })
+  @Path("profiles/{profile}")
+  public String getProfile(@PathParam("profile") String profileName) {
+    final GatewayConfig config = (GatewayConfig) request.getServletContext().getAttribute(GatewayConfig.GATEWAY_CONFIG_ATTRIBUTE);
+    final Map<String, Collection<String>> configuredProfiles = config.getHomePageProfiles();
+    if (configuredProfiles.containsKey(profileName)) {
```
[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/KnoxMetadataResource.java:262]: I'm going to make it case-insensitive.

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java]: Fixed.

File: gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/KnoxMetadataResource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (smolnar82) [reply gateway-service-metadata/src/main/java/org/apache/knox/gateway/service/metadata/KnoxMetadataResource.java]: Fixed.

File: gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java (hunk 1 above)
Comment (smolnar82) [reply gateway-server/src/main/java/org/apache/knox/gateway/config/impl/GatewayConfigImpl.java]: Fixed.

**Pull request merged**